inception-imls-stringmatch/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/stringmatch/StringMatchingRecommender.java

```diff
@@ -247,74 +243,34 @@ public EvaluationResult evaluate(List<CAS> aCasses, DataSplitter aDataSplitter)
         }
 
         // Predict
-        List<Sample> actualData = new ArrayList<>();
+        List<AnnotatedTokenPair> predictions = new ArrayList<>();
         for (Sample sample : testSet) {
-            List<Span> spans = new ArrayList<>();
-            
+
             for (TokenSpan token : sample.getTokens()) {
                 Trie<DictEntry>.Node node = dict.getNode(sample.getText(),
                         token.getBegin() - sample.getBegin());
-                if (node != null) {
-                    int begin = token.getBegin();
-                    int end = token.getBegin() + node.level;
-                    
-                    // Need to check that the match actually ends at a token boundary!
-                    if (sample.hasTokenEndingAt(end)) {
-                        for (LabelStats lc : node.value.getBest(maxRecommendations)) {
-                            spans.add(new Span(begin, end,
-                                    sample.getText().substring(begin - sample.getBegin(),
-                                            end - sample.getBegin()),
-                                    lc.getLabel(), lc.getRelFreq()));
-                        }
-                    }
+                int begin = token.getBegin();
+                int end = token.getEnd();
+
+                String predictedLabel = NO_LABEL;
+                if (node != null && sample.hasTokenEndingAt(token.getBegin() + node.level)) {
+                    List<LabelStats> labelStats = node.value.getBest(1);
+                    if (!labelStats.isEmpty())
+                        predictedLabel = labelStats.get(0).getLabel();
+                }
+                Optional<Span> coveringSpan = sample.getCoveringSpan(begin, end);
+                // check there is a gold label here
+                if (coveringSpan.isPresent()) {
+                    String goldLabel = coveringSpan.get().getLabel();
```

Comment:
I believe this is missing cases where there was a prediction but there is actually no gold label.

Reply:
The problem is that we don't know at this time if there will be a gold label in the future, it just hasn't been annotated yet. We can't really be sure that this type of error is an error. Therefore, I decided to count errors where we have a gold label but predicted NO_LABEL, but not these errors where we do not have a gold label (or a NO_LABEL instead). But you are right, there is also a case for counting these errors or even ignoring all cases without a label (or a NO_LABEL).

Reply:
What do you think about creating Pairs also for the case where there is no gold label and filter these cases out later when calculating the scores? That would allow us to handle this special case centrally when building the confusion matrix instead of having to consider it in every recommender implementation.

Reply:
That's a good idea. I will work on this.